backend/src/mail/mail.service.ts

```diff
@@ -164,14 +195,126 @@ export class MailService {
 
-    mail.contexts = mailContext.filter((itm) => isValid(itm.to[0]));
+    this.logger.log(
+      `Total emails to be sent (after filter null): ${mailContext.length}`,
+    );
+
+    mail.contexts = mailContext.filter((itm) => isValid(itm?.to[0]));
+
+    this.logger.log(
+      `Total emails to be sent (after filter invalid): ${mail.contexts?.length}`,
+    );
+
+    const invalidEmails = mailContext.filter((itm) => !isValid(itm.to[0]));
+
+    this.logger.log(`Total invalid emails filtered: ${invalidEmails?.length}`);
+
+    const mailQB = this.mailRepository
+      .createQueryBuilder('mail')
+      .leftJoinAndSelect('mail.tx', 'tx');
+    mailQB
+      .where('mail.email IN (:...emails)', {
+        emails: mail.contexts.map((itm) => itm.to[0]),
+      })
+      .andWhere('tx.tag = :tag', { tag: mail.contexts[0].tag })
+      .andWhere('mail.date > :date', {
+        date: new Date(new Date().setDate(new Date().getDate() - 6)),
+      });
+    if (
+      [EmailTags.MEMBER_FOLLOW_UP, EmailTags.SUPERVISOR_REMINDER].includes(
+        templateType,
+      )
+    ) {
+      this.logger.log(
+        'Filtering existing emails for automated notifications in the last 6 days',
+      );
+      const existingMail = await mailQB.getMany();
+
+      const existingEmails = existingMail?.map((itm) => itm.email);
+
+      this.logger.log(
+        `Total existing emails to filter: ${existingEmails?.length}`,
+      );
+
+      const filteredContexts = mail.contexts.filter(
+        (itm) => !existingEmails.includes(itm.to[0]),
+      );
+
+      if (existingEmails.length > 0) {
+        mail.contexts = filteredContexts;
+      }
+
+      this.logger.log(
+        `Total emails to be sent (after filter existing): ${mail.contexts?.length}`,
+      );
+
+      if (mail.contexts.length === 0) {
+        this.logger.log('No valid emails to send');
+        return;
+      }
+    }
 
     try {
       const res = await this.mailApi.post('/emailMerge', mail);
-      this.logger.log(res.data);
+
+      this.logger.log(
+        `Emails sent for ${templateType}: ${res.data.messages.length}`,
+      );
+
+      if (!res.data.txId) {
+        this.logger.error('No txId returned');
+        this.logger.log(res?.data);
+        return;
+      }
+
+      this.logger.log(res?.data);
+
+      const batch =
+        res.data.txId &&
+        (await this.mailBatchRepository.save(
+          this.mailBatchRepository.create({
+            txId: res.data.txId,
+            tag: mail.contexts[0].tag,
+            template: templateType,
+          }),
+        ));
+
+      const mails =
+        batch &&
+        res.data.messages.map((msg) => {
+          return this.mailRepository.create({
+            email: msg.to[0],
+            msgId: msg.msgId,
+            sent: true,
+            txId: batch.txId,
+            tx: batch,
+          });
+        });
+
+      mails.length > 0 && (await this.mailRepository.save(mails));
```

Comment:
By and large, I'd prefer using if statements to && just because they're more self explanatory. 

Also, in this case, mails **could** technically be empty from line 289 if, for example, `batch` is undefined for any reason, so this could crash when `mails.length` is searching for length of undefined